crates/optimism/consensus/src/proof.rs

```diff
@@ -0,0 +1,332 @@
+//! Helper function for Receipt root calculation for Optimism hardforks.
+
+use reth_primitives::{
+    ReceiptWithBloom,B256,
+};
+use reth_trie_common::root:: ordered_trie_root_with_encoder;
+use reth_chainspec:: {ChainSpec, OptimismHardfork};
+
+/// Calculates the receipt root for a header.
+pub fn calculate_receipt_root_optimism(
+    receipts: &[ReceiptWithBloom],
+    chain_spec: &ChainSpec,
+    timestamp: u64,
+) -> B256 {
+    // There is a minor bug in op-geth and op-erigon where in the Regolith hardfork,
+    // the receipt root calculation does not include the deposit nonce in the receipt
+    // encoding. In the Regolith Hardfork, we must strip the deposit nonce from the
+    // receipts before calculating the receipt root. This was corrected in the Canyon
+    // hardfork.
+    if chain_spec.is_fork_active_at_timestamp(OptimismHardfork::Regolith, timestamp) &&
+        !chain_spec
+            .is_fork_active_at_timestamp(OptimismHardfork::Canyon, timestamp)
+    {
+        let receipts = receipts
+            .iter()
+            .cloned()
+            .map(|mut r| {
+                r.receipt.deposit_nonce = None;
+                r
+            })
+            .collect::<Vec<_>>();
+
+        return ordered_trie_root_with_encoder(receipts.as_slice(), |r, buf| {
+            r.encode_inner(buf, false)
+        })
+    }
+
+    ordered_trie_root_with_encoder(receipts, |r, buf| r.encode_inner(buf, false))
+}
+
+#[cfg(feature = "optimism")]
```

Comment:
```suggestion
```

this feature is applied at crate level in _lib.rs_, so no need to put it in the rest of the code in this crate